proto/src/rr/domain/label.rs

```diff
@@ -0,0 +1,307 @@
+//! Labels are used as the internal components of a Name.
+//!
+//! A label is stored internally as ascii, where all unicode characters are converted to punycode interenally.
+
+#[allow(unused)]
+use std::ascii::AsciiExt;
+use std::cmp::{Ordering, PartialEq};
+use std::borrow::Borrow;
+use std::fmt::{self, Debug, Display, Formatter, Write};
+use std::hash::{Hash, Hasher};
+use std::str::FromStr;
+use std::sync::Arc as Rc;
+
+use idna::uts46;
+
+use error::*;
+
+const WILDCARD: &[u8] = b"*";
+const IDNA_PREFIX: &[u8] = b"xn--";
+
+/// Labels are always stored as ASCII, unicode characters must be encoded with punycode
+#[derive(Clone, Eq)]
+pub struct Label(Rc<[u8]>);
+
+impl Label {
+    /// These must only be ASCII, with unicode encoded to PunyCode
+    ///
+    /// This uses the bytes as raw ascii values.
+    pub fn from_bytes(bytes: &[u8]) -> Result<Self, ProtoError> {
+        Ok(Label(Rc::from(bytes)))
+    }
+
+    /// Converts this label to lowercase
+    pub fn to_lowercase(&self) -> Self {
+        // TODO: replace case conversion when (ascii_ctype #39658) stabilizes
+        if let Some((idx, _)) = self.0.iter().enumerate().find(|&(_, c)| *c != c.to_ascii_lowercase()) {
+            let mut lower_label: Vec<u8> = self.0.to_vec();
+            lower_label[idx..].make_ascii_lowercase();
+            Label(Rc::from(lower_label))
+        } else {
+            self.clone()
+        }
+    }
+
+    /// Returns true if this label is the wildcard, '*', label
+    pub fn is_wildcard(&self) -> bool {
+        self.as_bytes() == WILDCARD
+    }
+
+    /// Returns the lenght in bytes of this label
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    /// Returns the raw bytes of the label, this is good for writing to the wire.
+    ///
+    /// See [`Display`] for presentation version (unescaped from punycode, etc)
+    pub fn as_bytes(&self) -> &[u8] {
+        &self.0
+    }
+
+    /// Performs the equivelence operation disregarding case
+    pub fn eq_ignore_ascii_case(&self, other: &Self) -> bool {
+        self.0.eq_ignore_ascii_case(&other.0)
+    }
+
+    /// compares with the other label, ignoring case
+    pub fn cmp_with_f<F: LabelCmp>(&self, other: &Self) -> Ordering {
+        let s = self.0.iter();
+        let o = other.0.iter();
+        
+        for (s, o) in s.zip(o) {
+            match F::cmp_u8(*s, *o) {
+                Ordering::Equal => continue,
+                not_eq => return not_eq,
+            }
+        }
+
+        self.0.len().cmp(&other.0.len())
+    }
+}
+
+impl AsRef<[u8]> for Label {
+    fn as_ref(&self) -> &[u8] {
+        &self.0
+    }
+}
+
+impl Borrow<[u8]> for Label {
+    fn borrow(&self) -> &[u8] {
+        &self.0
+    }
+}
+
+impl FromStr for Label {
+    type Err = ProtoError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
```

Comment:
This is not a good name because it doesn't give an indication of what the input encoding is supposed to be.

Reply:
Sorry, are you saying Label isn't a good name? or that we should not implement FromStr on Label?

Reply:
> Sorry, are you saying Label isn't a good name? or that we should not implement FromStr on Label?

I'm saying from_str is ambiguous because it doesn't indicate how the input string will be interpretered; i..e. what encoding it is expected to be in.

Reply:
ok, answered in the above comment.